src/Validation/URLValidationCron.php

```diff
@@ -0,0 +1,142 @@
+<?php
+/**
+ * WP cron process to validate URLs in the background.
+ *
+ * @package AMP
+ * @since 2.1
+ */
+
+namespace AmpProject\AmpWP\Validation;
+
+use AmpProject\AmpWP\BackgroundTask\BackgroundTaskDeactivator;
+use AmpProject\AmpWP\BackgroundTask\RecurringBackgroundTask;
+use AmpProject\AmpWP\Infrastructure\Injector;
+
+/**
+ * URLValidationCron class.
+ *
+ * @since 2.1
+ *
+ * @internal
+ */
+final class URLValidationCron extends RecurringBackgroundTask {
+
+	/**
+	 * Injector instance.
+	 *
+	 * @var Injector
+	 */
+	private $injector;
+
+	/**
+	 * The cron action name.
+	 *
+	 * @var string
+	 */
+	const BACKGROUND_TASK_NAME = 'amp_validate_urls';
+
+	/**
+	 * The number of URLs to check per type each time the cron action runs.
+	 *
+	 * @var int
+	 */
+	const DEFAULT_LIMIT_PER_TYPE = 1;
+
+	/**
+	 * The length of time, in seconds, to sleep between each URL validation.
+	 *
+	 * @var int
+	 */
+	const DEFAULT_SLEEP_TIME = 1;
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param BackgroundTaskDeactivator $background_task_deactivator Service that deactivates background events.
+	 * @param Injector                  $injector Injector instance.
+	 */
+	public function __construct( BackgroundTaskDeactivator $background_task_deactivator, Injector $injector ) {
```

Comment:
Both of these classes are not necessarily shared services. Multiple instances with different options may be created in different places within a single execution (although I don't think that we're doing that anywhere right now outside testing). What's the recommended approach for creating non-shared instances (potentially with different options passed into their constructors) with our dependency injection?

Reply:
Nothing is shared by the injector by default, every injection gets its own separate instance.

If you want to have an instance shared across all injections, it needs to be added to the list of shared instances:

https://github.com/ampproject/amp-wp/blob/2.0.8/src/AmpWpPlugin.php#L145-L165

Reply:
So, in this case, just inject the services where needed, and it will just work, no need for injecting the injector.

Reply:
All right, so if I add `ScannableURLProvider` to the constructor here, how would I pass arguments into `ScannableURLProvider`'s constructor when it's instantiated?

Reply:
Like this? `$this->injector->make( ScannableURLProvider::class, [ /* arguments */ ] )`?

Reply:
@westonruter Alain is suggesting that I not use the inject directly but instead declare `ScannableURLProvider` as a dependency in the constructor. So what I'm unsure of is, if I have,

```
public function __construct( BackgroundTaskDeactivator $background_task_deactivator, ScannableURLProvider $scannable_url_provider ) {
```

how do I pass arguments into `$scannable_url_provider`'s constructor?

I'll dig into the dependency injection code to see if there's an answer built in.

Reply:
There are a few ways to solve this.
Let's summarize the constraints first (to make sure I understand correctly):
- `ScannableURLProvider` needs 3 arguments that differ between instantiations.
- These arguments need to be decided "late" as they depend on a filter, and we don't want to call that filter too early.

a. Inject a `ScannableURLProviderFactory` instead of a `ScannableURLProvider`. So, instead of injecting the entire Injector (which gives too much knowledge and power to the object that receives it, and couples it to knowledge about technical implementation details that are irrelevant to its business domain), we inject the possibility to decide on what and when to instantiate regarding the `ScannableURLProvider` only.

b. Turn the constructor arguments into an injectable object, like `URLScanningContext`. This makes it controllable again by the injector, and the injector can then include logic that decides what to inject. In this case, though, it is enough to just not have any extra logic in the injector at all. All we need is to lazily provide the needed settings only at the point where they are actually requested. The `URLScanningContext` could at that point run the filter transparently.

c. Move the constructor arguments from the constructor into the public method that triggers the action. So, instead of having 3 constructor arguments and 1 method arguments, you're left with 4 method arguments.

All 3 of the above should work. I think my preference is with b., as it cleanly separates the logic into meaningful bits and deals with the timing issue (which happens to be a WordPress technical detail) in a transparent way.

Feel free to ask questions about the above, or let me know if you want to chat in a call to discuss in more detail.

Reply:
Hmm, looking back at the code, I think b. would not solve the third constraint I failed to recognize:
- The `URLValidationCron` wants to control the 3 constructor arguments when it runs and control the knowledge about them.

I think this requirement is not necessarily one I'd enforce, as it seems that has nothing to do with a cron job, but is more about resource allocation of the background scanning. So, I'd be fine with moving `get_url_validation_number_per_type` into the `URLScanningContext` instead. However, if you want to avoid that, you can go with a. above instead.

Reply:
I went with option b and added `URLScanningContext` in 6747c04f03f0d480d71567ab6ddf165e9375eecf. 

I think in the future it may be necessary to add some setters to `URLScanningContext` to allow changing the configuration later in the process. For now the defaults work for our use cases, except in the CLI script, which is still outside the DI architecture so creates its own instances on the fly.